### DISCUSSION TITLE
autodiff hessian of betaln function is wrong

### DISCUSSION THREAD

**Discussion by epenn-nx** [jax-ml/jax#34214] [Q&A]
I am aware that 'wrong' is a strong word here.  In the example below, when I sample or plot the function, the curvature is uniformly positive, but when I use the autodiff i get a negative number. I have also confirmed that running `grad` twice in the appropriate way leads to the same outcome. 

Am i missing something?  Is there a workaround?  I'm using this in a 2nd order optimization and running into some issues.

## MRE

```python
import jax
import jax.numpy as jnp
from jax.scipy.special import betaln

# Test point: α = β = 1 (in log-space: θ_α = θ_β = 0)
log_alpha = 0.0
log_beta = 0.0

# Loss function using betaln
def loss_func(theta_a, theta_b):
    a = jnp.exp(theta_a)
    b = jnp.exp(theta_b)
    return betaln(a, b) - betaln(a + 1, b)

# Compute Hessian via autodiff
hess_autodiff = jax.hessian(loss_func, argnums=(0,1))(log_alpha, log_beta)
print(f"Hessian (autodiff): ({hess_autodiff[0][0]:.6f}, {hess_autodiff[1][1]:.6f})")

# Compute Hessian via numerical finite differences
h = 0.01
loss_center = float(loss_func(log_alpha, log_beta))
loss_plus_a = float(loss_func(log_alpha + h, log_beta))
loss_minus_a = float(loss_func(log_alpha - h, log_beta))
loss_plus_b = float(loss_func(log_alpha, log_beta + h))
loss_minus_b = float(loss_func(log_alpha, log_beta - h))

hess_aa_numerical = (loss_plus_a - 2*loss_center + loss_minus_a) / (h**2)
hess_bb_numerical = (loss_plus_b - 2*loss_center + loss_minus_b) / (h**2)
print(f"Hessian (numerical): ({hess_aa_numerical:.6f}, {hess_bb_numerical:.6f})")```
```

Output:
```
Hessian (autodiff):   (-0.572467, -0.572467) 
Hessian (numerical):  ( 0.231266,  0.255108)  
```

## Environment

- **JAX version**: 0.6.2
- **jaxlib version**: 0.6.2  
- **Python**: 3.10
- **OS**: Linux

**Comment by DiagRisker** [comment]
Hello @epenn-nx ,

I will share some code to do manually the Hessian and higher derivatives automatically (so that you can decompose higher step if needed).
On a quick sight, playing with the `h` parameter, your differentiation is not very much stable (expectable in log space though):
```python
for h in [1e-1,1e-2,1e-3,1e-4,1e-5]:
    print('h = ',h)
    loss_center = float(loss_func(log_alpha, log_beta))
    loss_plus_a = float(loss_func(log_alpha + h, log_beta))
    loss_minus_a = float(loss_func(log_alpha - h, log_beta))
    loss_plus_b = float(loss_func(log_alpha, log_beta + h))
    loss_minus_b = float(loss_func(log_alpha, log_beta - h))

    hess_aa_numerical = (loss_plus_a - 2*loss_center + loss_minus_a) / (h**2)
    hess_bb_numerical = (loss_plus_b - 2*loss_center + loss_minus_b) / (h**2)
    print(f"Hessian (numerical): ({hess_aa_numerical:.6f}, {hess_bb_numerical:.6f})")
```

> h =  0.1
> Hessian (numerical): (0.249815, 0.249958)
> h =  0.01
> Hessian (numerical): (0.231266, 0.255108)
> h =  0.001
> Hessian (numerical): (0.000000, 0.953674)
> h =  0.0001
> Hessian (numerical): (-143.051147, 0.000000)
> h =  1e-05
> Hessian (numerical): (-21457.672119, -2384.185791)

It might be the numerical differentiating step that is not right

EDIT (the code for each step) _____
Say that I do manually each steps (i included some personal code already shown in https://github.com/jax-ml/jax/issues/32194#event-20071503280 ):
```python
def mvmap(f: Callable, ni: int, na : int = 0):
    """ multi vmap wrapper for per component arithmetics and more (if ni < input.ndim)
    for Chain Gradient, this will have a right tensor product behavior (more efficient for divergence application )"""
    if not ni-1: return jax.vmap(f, in_axes = (0,)*na if na else 0, out_axes = (0,)*na if na else 0); 
    return mvmap(jax.vmap(f, in_axes = (0,)*na if na else 0, out_axes = (0,)*na if na else 0), ni-1) 


def D(f: Callable):
    """ Diff operator """
    def vj(*x):
        y, pullback = jax.vjp(f, *x)
        if hasattr(y, "ndim"):
            if not y.ndim:
                return pullback( 1.)
            else:
                return mvmap(pullback, ni = y.ndim 
                )( otimes(*(jnp.ones(y.shape),)*2) ) 
        if not sum( jax.tree_util.tree_map(lambda a : a.ndim, y) ):
            return pullback( tuple(1. for o in y))
        return mvmap(pullback, ni = y[0].ndim, na = len(y) 
        )( tuple( jnp.ones(o.shape*2) if o.ndim else 1. for o in y) )
    return vj 

def RS(f: Callable, n: int): 
    """ recursive part """
    if not n-1: # if n == 1
        return D(f) 
    return RS( D(f), n-1) 


Dlfunc = jax.tree_util.tree_map(D,loss_func)
print( " 1st order derivative : ", Dlfunc(log_alpha, log_beta) )
D2lfunc = RS(loss_func, 2) # same as jax.tree_util.tree_map(D,Dlfunc)
print( " 2nd order derivative : ", D2lfunc(log_alpha, log_beta))
```
I don't obtain the same values (it's inteded to pick the derivative per variable (2nd order), so it is still the same 2 diagonal values of the Hessian you extract, cf hess_autodiff[0][0], hess_autodiff[1][1] )
